### PR TITLE
Implement key-value storage

### PR DESCRIPTION
**Pull request by MytsV**
# Purpose

Quite evidently, to store data between program runs.

# Summary

Created an implementation of a very simple storage system. Its interface consists only of `getAll`, `add`, `edit` and `delete` methods. All of them take entity type as an argument.

Each type of entity is saved in a separate file in the database directory. The content of it is encoded with Vigenere cipher (**NOT FOR DATA SECURITY!** Just to discourage users from changing the files by hand).

Content format is extremely simple: first two bytes denote a uint16 number, indicating the last id written to the database. Therefore, entity's unique id is implemented with a 0 <= number < 2^16. The following bytes are an UTF-8 encoded string of JSON data, which is easily parsed by built-in NodeJs functions.

# Other information

The problem with this implementation is finite and quite limited number of IDs. It will be addressed later by creating a `prune` method, which will reindex the database.

### CODE REVIEW

File: src/cipher.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  return result;
+};
+
+const encode = (bytes) => {
+  return rotate(bytes);
+};
+
+const decode = (bytes) => {
+  return rotate(bytes, -1);
+};
```
Comment (svitbar) [inline src/cipher.js]: I'd rather rewrite it like this. But it depends on you)
```suggestion
const encode = (bytes) => rotate(bytes);

const decode = (bytes) => rotate(bytes, -1);
```

File: src/database.js [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+const path = require('path');
+
+// ID length in bytes
+const idSize = 2;
+
+const toBytes = (data) => {
+  const idBuffer = Buffer.alloc(idSize);
+  idBuffer.writeUInt16BE(data.lastIdx);
+  // Write json string with default UTF-8 encoding
+  const jsonBuffer = Buffer.from(JSON.stringify(data.entities));
```
Comment (svitbar) [inline src/database.js]: I'd rather divide this to enhance readability.
```suggestion
  const jsonStr = JSON.stringify(data.entities);
  
  const jsonBuffer = Buffer.from(jsonStr);
```

File: src/database.js [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+const isEntityCorrect = (entity) => {
+  for (const value of Object.values(entity)) {
+    if (typeof value === 'object') {
+      if (!isEntityCorrect(value)) return false;
+    } else if (typeof value !== 'string' && typeof value !== 'number') {
+      return false;
+    }
+  }
+  return true;
+};
```
Comment (svitbar) [inline src/database.js:52]: I don't really like this 'if-else' structure, but not sure that it is possible to rewrite it in a better way.

File: src/database.js [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+/** Implementation of single-file entity{key-value} storage */
+class SimpleDatabase {
+  /**
```
Comment (svitbar) [inline src/database.js:56]: I'd rather take this class out to a separate file.

**Review by svitbar — CHANGES_REQUESTED** [review]
All seems to be great. Thank you so much.
I added just a few suggestions.

File: src/cipher.js (hunk 1 above)
Comment (MytsV) [reply src/cipher.js]: Thanks, will rewrite it this way

File: src/database.js (hunk 2 above)
Comment (MytsV) [reply src/database.js]: Agree with you, though I don't really like the jsonStr name

File: src/database.js (hunk 3 above)
Comment (MytsV) [reply src/database.js:52]: It can be some forEach chained atrocity, you wouldn't prefer it :)

File: src/database.js (hunk 4 above)
Comment (MytsV) [reply src/database.js:56]: But there are lots of helper functions, I don't want to export them anywhere

File: src/database.js (hunk 2 above)
Comment (svitbar) [reply src/database.js]: You can rename it as well. It was just the first thing that came up to my mind)

File: src/database.js (hunk 3 above)
Comment (svitbar) [reply src/database.js:52]: Then better to leave as it is.

**Review by svitbar — APPROVED** [review]
I agree with your comments.

**Pull request merged**